### ISSUE TITLE
Blazer: LEFT OUTER JOIN query never finishes

### ISSUE THREAD

**Issue by tripleee** [Charcoal-SE/metasmoke#921]
I created a query which simply ends up displaying the browser's "Aw, snap" sad face eventually. I have had experiences in the past where Blazer queries were slow and timed out, but this seems like a different beast, as the timeout is not caught and reported by Ruby or other parts of the web framework. I'm wondering if I managed to DoS metasmoke somehow.

The problematic query is https://metasmoke.erwaysoftware.com/data/sql/queries/297-phone-number-in-title-is-not-blacklisted

**Comment by tripleee** [comment]
A variation which avoids the (allegedly more efficient) LEFT OUTER join https://metasmoke.erwaysoftware.com/data/sql/queries/298-phone-number-in-title-is-not-blacklisted finishes quickly and returns some 3,000 rows.

**Comment by makyen** [comment]
You've given us nothing but a link to the query. Queries are *automatically run* when you go to the page. You say you're concerned about DoS'ing MS as a result of this query. So, you've left us little alternative for investigation other than clicking on the link you've provided and contributing to the DoS ourselves, if that's what's happening.

**Comment by tripleee** [comment]
Sorry about that. Here is the query text:

```
-- https://stackoverflow.com/a/14710831/874188
SELECT p.id, p.title, p.body FROM p_posts AS p
LEFT OUTER JOIN p_posts_reasons AS r
             ON r.post_id = p.id
            AND r.reason_id in (47, 164)
            -- 47 = phone number in title
            -- 164 = bad phone number in title
WHERE r.post_id IS NULL
```

Here is a query to fetch the information about an individual query by its Blazer ID:

https://metasmoke.erwaysoftware.com/data/sql/queries/299-blazer-query-by-id?query_id=297

**Comment by makyen** [comment]
I did actually click on the link you provided. Your query was cached about an hour ago. It returned 357,465 rows. I suspect that you didn't crash MS, but that you did crash the browser tab you were using when it tried to display all 357,465 rows of results.

When experimenting/working on a query, I've found it's a good idea to include a `LIMIT` to get a short look at if the results resemble something you want. When removing the `LIMIT` when your query is still exceeding the limit you've set, it's a good idea to run it with a `COUNT` instead of the unlimited query to get a feel for the number of results which you're returning. I've edited the query to have `LIMIT 100`.



Here's an image of the prior results:
[![enter image description here][1]][1]

  [1]: https://i.stack.imgur.com/wgvzB.png
